### ISSUE TITLE
Concurrent access to Event Loop?

### ISSUE THREAD

**Issue by polyfractal** [carllerche/mio#163]
I posed this question on IRC earlier, but thought it might be easier to articulate myself on an issue.  Preface to say that I've never actually worked with kqueue / epoll before, so this may all be wildly wrong or terrible :)

I'm curious if the EventLoop could be made to be Sync, or Clone-able so that multiple threads can access it concurrently.  Or perhaps a specialized OneShotEventLoop that only allows OneShots but in exchange is Sync or Clone-able?

My motivation is to register all events with OneShot so that multiple threads can work off the event loop like a work queue. If I understand correctly, kqueue / epoll is safe to use concurrently if events are registered with OneShot, since the FD is de-armed as soon as the first thread accesses it.

The single-threaded alternative is to have a one thread spinning the event loop and delegating "work" to other threads.  This isn't ideal, since any message passing boils down to some kind of concurrent queue which will cause extra coherence traffic between cores.  It is also poor for cache usage, since the event loop thread has the data hot in L1 / L2 cache, but offloads it to another thread which is probably on a different core.  That cold core needs to fetch the data and populate it's own L1 / L2.

Instead, I would like multiple threads each spinning the event loop concurrently, pulling down OneShot "work units".  This has the advantage of not requiring coordination between threads, and data pulled in one thread stays local so it utilizes caches better.  Each thread would then move through a cycle of "produce local work -> consume local work", rather than a single producing thread feeding multiple worker threads.

Now, that's the idea...but it may be ridiculous from an kqueue / epoll point of view.  Thoughts?  You all know a lot more about this than myself :)

**Comment by rrichardson** [comment]
First and foremost, polling APIs (kqueue, epoll, etc) do one thing: They wait block while waiting for something interesting to happen to an array of file descriptors.  That is the poll() functionality.  This, necessarily can only happen in a single thread.  Once its passed a timeout or events have occurred, the results of the activity are returned in a new array.  

See:  https://github.com/carllerche/mio/blob/master/src/os/epoll.rs#L22-L32 

So we construct an array of events in which we're interested (line 22)  then we wait for timeout_ms or less for something to happen.  Then the resulting fds that had an event fire will be present in the array after line 29.  

While this is polling, changing the array has undefined results, that is to say, you shouldn't do it. So concurrently updating the poll list is no bueno. That's why almost all APIs around epoll make this operate in a single threaded loop, where it collects things of interest, polls them, then notifies listeners of the events. 

This is the case regardless of whether the events are edge/level/one-shot, etc. 

From this point, you have an array of file descriptors, **you can then pass these fds to a thread pool** so that they can do whatever work they need re: sockets and whatever.

Just so we're clear, the thing on which kqueue and epoll operate is file descriptors, not work-units. It is not a scheduler, though it can be used that way, but it only schedules the management of inbound or outbound IO.

## on performance

Somewhere you're going to have to pay the tax for marshalling access to multiple threads. 

Somewhere there is a queue, either feeding fds into worker threads, or reading buffers and feeding buffer handles into worker threads.  

If you actually do want cache coherency you need some way to ensure that the sockets are consistently managed by the same threads/cores, so you need some sort of consistent sharding work pool.  Also, you need some efficient way to re-register events from the work you've done in these threads, which means you still need to to be able to register interest in those events via a queue.  You could use a mutex to ensure that you're not concurrently updating the file-descriptor array while epoll is polling, but I wouldn't recommend it.  One reason is epoll is guaranteed to be blocking for a relatively long amount of time, so that is guaranteed to block all of your threads that need to register for a relatively long amount of time. 

So this means that Mio is already set up to do exactly what you need.   Basically you can set up a Slab or some sort of Token->connection mapping to get file descriptors and connection metadata from each event.  You have each work-pool loop have one of these.  

Then in the Mio Handler,  you dole the work out using a hashing scheme to each thread pool.  Then each worker registers events in the queue by sending requests back in through the notify() queue. 

## The big picture 

So there is overhead in managing a work pool with kqueue,  if you're looking for a work scheduler, this is definitely not it, for that you should use some sort of work-stealing queue, like a concurrent circular list or something.

**Comment by carllerche** [comment]
I believe the tl;dr is that `EventLoop` should not be `Sync` and to ideal pattern would be to have 1 `EventLoop` instance per thread :smile:

Going to close this ticket. Feel free to open others.

**Comment by ayosec** [comment]
> So we construct an array of events in which we're interested (line 22) then we wait for timeout_ms or less for something to happen. [...]
>
> While this is polling, changing the array has undefined results, that is to say, you shouldn't do it.

Maybe I'm wrong, but this seems to be a limitation of the current implementation in mio. I don't known about kqueue, but with epoll it is safe to access to the same queue from multiple threads if you add file descriptions with `EPOLLONESHOT`.

**Comment by rrichardson** [comment]
> I don't known about kqueue, but with epoll it is safe to access to the same queue from multiple threads if you add file descriptions with EPOLLONESHOT

What is the queue?  There isn't really a queue in epoll/kqueue

**Comment by ayosec** [comment]
> What is the queue? There isn't really a queue in epoll/kqueue

My bad. I mean the epoll/kqueue instance.

**Comment by rrichardson** [comment]
Where is this information posted? Please don't say Stack Overflow. 

I still stand by my previous statement: while epoll is polling, you are not allowed to change the array of fds which it is polling.  If you want to coordinate the two phases, 1. setting the array, and then 2. waiting on the poll(), you can, but that means you have to coordinate this all with a Mutex or something worse. Now all of your threads are moving in lock-step and waiting on poll() together, which is completely destroying your whole point concurrency in the first place.

Can/can't might be the wrong set of questions here,  maybe should/shouldn't is better.

**Comment by mikedilger** [comment]
> That is the poll() functionality. This, necessarily can only happen in a single thread
...
> while epoll is polling, you are not allowed to change the array of fds which it is polling

I know of no such linux limitation.  Please provide a reference @rrichardson 

Many developers use epoll_wait(epfd) on the same epfd in multiple threads, using EPOLLONESHOT, including (for example) Jason Baron at Akamai, and the GlusterFS folks AFAIK.

In edge-triggered mode, just 1 thread is woken up to handle the event.  The kernel picks one.  In level-triggered mode, they all wake up, but EPOLLONESHOT prevents all but one from actually getting the event.   Jason Baron's "thundering herd" patch is meant to solve this excessive thread awakening.

There is AFAIK still no threadsafe way to remove an fd from the epoll instance, and remove associated data shared among the worker threads.  EPOLL_CTL_DISABLE (proposed for 3.7, but reverted) would have fixed that.
* EPOLL_CTL_DISABLE original thread: http://thread.gmane.org/gmane.linux.kernel/1311457
* EPOLL_CTL_DISABLE article (was reverted):  https://lwn.net/Articles/520012/

**Comment by rrichardson** [comment]
I should have specified that there is no safe way to remove from the array,
adding should be easier, but may not result in the intended behavior on the
first poll.

That said, I was unaware of the work that Jason Baron had done, thanks for
pointing that out.

It would be tricky to make an abstraction that works for platforms other
than Linux, but I could imagine a sync'able inner event loop object which
could be exposed that is somewhat like a RWLock, allowing register and
reregister functions all the time, and deregister functions when not
poll'ing. If we got clever we could even make the deregister function block
waiting on the poll with thread parking.

Can you point to an actual code example of multiple threads waiting on the
same epfd?

**Comment by rrichardson** [comment]
Another option, which would be easier to abstract over, is build the thread
pool/scheduling functionality into the event loop. That would give us more
latitude to abstract over the different APIs.

**Comment by mikedilger** [comment]
> Can you point to an actual code example of multiple threads waiting on the same epfd?

I wrote one a few years back.  This clearly is not evidence that it is a good idea, though.
https://github.com/mikedilger/epoll_demo (warning: it's not using EPOLLONESHOT  / EPOLL_CTL_MOD, so events might be handled out of order)

I just checked nginx, and its threads pull tasks off of a task queue which is populated by the thread waiting on the epoll.

I think most developers feel that "epoll's multithread behavior is too uncertain" to use it the way I (and others above) have suggested.

> Another option, which would be easier to abstract over, is build the thread pool/scheduling functionality into the event loop. That would give us more latitude to abstract over the different APIs.

That is exactly what I was thinking about.

That is why I found this month old closed bug.  I was looking to write a thread pool library on top of mio, and when I found the epoll usage was not as I had envisioned, I searched for issues from other people who may have noticed the same thing, and found this one.

IMHO, mio should continue on as it is.  Using epoll in a multithreaded way will probably never be quite safe.  And the readme says thread pools are an explicit non-goal of mio.   Perhaps eventually some sort of mt-epoll-tp library could expose the same interface as a thread pool library written on mio.

**Comment by rrichardson** [comment]
I was thinking about this a bit more, and, actually there are two "control"
interfaces to EventLoop,  The "external" interface that it exposes to the
owner of the object, and the interface that it offers to the callback
handlers. In EventLoop they are the same, but they don't have to be (I make
a distinction in my wrapper around EventLoop here:
https://github.com/rrichardson/reactor)

The interesting guarantee that is made by  the callback/internal interface
is that it is guaranteed to never be exposed while the poller is polling.
This means we could safely put functions like deregister in there.  For the
external interface, functions like register would either operate directly
(in the case of epoll)  or push into a thread-safe queue to be applied
inbetween poll() (in the case of kqueue)

I guess what I'm proposing is a fairly simple way to make a Sync outer
interface without punishing the users who only use EventLoop in a
single-threaded model. It would take some gymnastics to offer parallelism
for kqueue/IOCP users, though.

**Comment by rrichardson** [comment]
And yea, to your point, it doesn't need to exist inside Mio, as long as we
can expose the appropriate bits from EventLoop
